File: Graph/Graph_Depth_First/GraphDFS.java
import java.util.Queue;
import java.util.LinkedList;

/**
 * Depth-First Search implementation class.
 * @author Yasin Radi
 */
public class GraphDFS {

    // Num of vertices
    private int numVertex;

    // Adjacency list 
    private LinkedList<Integer>[] adj;

    // Constructor
    public GraphDFS(int n) {
        this.numVertex = n;
        this.adj = (LinkedList<Integer>[]) new LinkedList<?>[n];
        for(int i = 0; i < n; i++)
            this.adj[i] = new LinkedList<Integer>();
    };

    /**
     * Recursive function to perform a DFS traversal.
     * @param nodeIndex     int
     * @param visited       boolean
     * @param sb            StringBuilder
     */
    private void DFSUtil(int nodeIndex, boolean[] visited, StringBuilder sb) {
        // Mark the current node as visited
        // And append it for printing
        visited[nodeIndex] = true;
        sb.append(nodeIndex).append(' ');

        // Recurr for all the vertices
        // adjacent to this vertex
        for(int element : this.adj[nodeIndex])
            if(!visited[element]) 
                DFSUtil(element, visited, sb);
    };

    /**
     * Adds a new edge to the Graph.
     * @param nodeIndex     int
     * @param newNodeValue  int
     */
    public void addEdge(int nodeIndex, int newNodeValue) {
        this.adj[nodeIndex].add(newNodeValue);
    };

    /**
     * Performs a Depth-First traversal to display
     * all nodes contained in Graph.
     * @param nodeIndex     int
    */
    public void DFS(int nodeIndex) {
        // Builder to display nodes content
        StringBuilder sb = new StringBuilder();

        // Mark all the vertices as not visited
        boolean[] visited = new boolean[this.numVertex];
        for(int i = 0; i < this.numVertex; i++) 
            visited[i] = false;

        // Call the recursive helper method to
        // display DFS traversal.
        DFSUtil(nodeIndex, visited, sb);

        System.out.println(sb.toString());
    };

    public static void main(String[] args) {
        GraphDFS g = new GraphDFS(4);

        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 2);
        g.addEdge(2, 0);
        g.addEdge(2, 3);
        g.addEdge(3, 3);

        System.out.println("DFS starting from v2: ");
        g.DFS(2);

        System.out.println("DFS starting from v0: ");
        g.DFS(0);

        System.out.println("DFS starting from v1: ");
        g.DFS(1);

        System.out.println("DFS starting from v3: ");
        g.DFS(3);
    };
};

File: Graph/Graph_Depth_First/GraphDFS.cs
using System;
using System.Collections.Generic;
using System.Text;


/// <summary>
/// Depth-First Search implementation class.
/// </summary>
public class GraphDFS
{
    // Num of vertices
    private int numVertex;

    // Adjacency list
    private LinkedList<int>[] adj;

    // Constructor
    public GraphDFS(int n) {
        this.numVertex = n;
        this.adj = new LinkedList<int>[n];
        for (int i = 0; i < n; i++)
            this.adj[i] = new LinkedList<int>();
    }

    /// <summary>
    /// Recursive function to perform a DFS traversal.
    /// </summary>
    /// <param name="nodeIndex">Index of the node to start with.</param>
    /// <param name="visited">Array holding node visited status.</param>
    /// <param name="sb">StringBuilder to perform value printing.</param>
    private void DFSUtil(int nodeIndex, ref bool[] visited, StringBuilder sb) {
        // Mark current node as visited
        // Append it for printing
        visited[nodeIndex] = true;
        sb.Append(nodeIndex).Append(" ");

        // Recurr for all vertices
        // adjacent to this vertex
        foreach (int element in adj[nodeIndex])
            if (!visited[element])
                DFSUtil(element, ref visited, sb);
    }

    /// <summary>
    /// Adds an edge to the graph.
    /// </summary>
    /// <param name="nodeIndex"> Index of node to add.</param>
    /// <param name="newNodeValue"> Value of the node being added.
    public void AddEdge(int nodeIndex, int newNodeValue) {
        adj[nodeIndex].AddLast(newNodeValue);
    }

    /// <summary>
    /// Performs a Depth-First traversal to display
    /// all nodes contained in Graph.
    /// </summary>
    /// <param name="nodeIndex">Index of the node to start with.</param>
    public void DFS(int nodeIndex) {
        // Builder to display node content
        StringBuilder sb = new StringBuilder();

        // Mark all vertices as not visited
        bool[] visited = new bool[numVertex];
        for (int i = 0; i < numVertex; i++)
            visited[i] = false;

        // Call recursive helper to display DFS traversal.
        DFSUtil(nodeIndex, ref visited, sb);

        Console.WriteLine(sb.ToString());
    }
}
